compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -617,6 +620,33 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     });
 
                 if let Some((decl, body_id)) = closure_decl_and_body_id {
+                    let mut visitor = UnresolvedTypeFinder::new(self);
+                    for arg in fn_sig.inputs().skip_binder().iter() {
+                        if let ControlFlow::Break((ty, _)) = visitor.visit_ty(*arg) {
```

Comment:
Can you use [`InferCtxt::unresolved_type_vars`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/infer/struct.InferCtxt.html#method.unresolved_type_vars) instead of invoking `UnresolvedTypeFinder` itself?

Also, you probably don't need to use `type_variables()` at all -- you can probably just take the span that's returned from `unresolved_type_vars` and do `.unwrap_or(ty_span)`.

The `UnresolvedTypeFinder` already tries to look up the `var_origin` of the `ty_vid` in question, so you're doing a bit of duplicated work below.

Reply:
r=me after this

Reply:
So for some reason, the span is always None when I do this (which is afaict basically the same as what I already have, note that I am currently ignoring the span). I can convert to this and then when it's `None` fallback to the current implementation?

Reply:
Presumably there needs to be another case here? https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_infer/infer/resolve.rs.html#139 I think that would probably be a separate PR but lmk if you want me to take a stab at it

Reply:
Oh, no actually you should do this: 

1.) Edit the `UnresolvedTypeFinder` implementation of `TypeVisitor` to return the `span` _always_, but also get it to return the `TypeVariableOriginKind` (the logic is [here](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_infer/infer/resolve.rs.html#139)) too. The `BreakTy` should then be `Ty, Span, TypeVariableOriginKind`. 

2.) Then edit the existing logic that uses `UnresolvedTypeFinder` to now be conditional (I think it's just used in `generator_interior.rs`) on that `TypeVariableOriginKind` being of type `TypeVariableOriginKind::TypeParameterDefinition`.

Reply:
Edit: actually, the signature will be `(Ty, Option<Span>, TypeVariableOriginKind)`. We still probably need to `unwrap_or_else` the span to deal with when it's `None`, but I think you should fall back to the outer span of the closure. You shouldn't need to inspect `infcx.inner` in your logic.